server/aesdsocket.c

```diff
@@ -11,59 +11,146 @@
 #include <fcntl.h>
 #include <syslog.h>
 #include <sys/stat.h>
+#include <pthread.h>
+#include <time.h>
+#include <sys/queue.h>
 
 #define PORT 9000
 #define BUFFER_SIZE 1024
 #define FILE_PATH "/var/tmp/aesdsocketdata"
 
 int server_socket = -1;
-int client_socket = -1;
+pthread_mutex_t file_mutex = PTHREAD_MUTEX_INITIALIZER;
 
+// Define structure for storing thread info
+typedef struct client_thread {
+    pthread_t thread_id;
+    int client_socket;
+    LIST_ENTRY(client_thread) entries;
+} client_thread_t;
+
+// Define head for linked list
+LIST_HEAD(thread_list, client_thread) thread_head;
+
+// Signal handler for cleanup
 void handle_signal(int sig) {
     syslog(LOG_INFO, "Caught signal, exiting");
-    if (client_socket != -1) close(client_socket);
+
+    // Close all active connections
+    client_thread_t *thread_data;
+    while (!LIST_EMPTY(&thread_head)) {
+        thread_data = LIST_FIRST(&thread_head);
+        pthread_cancel(thread_data->thread_id);
+        pthread_join(thread_data->thread_id, NULL);
+        LIST_REMOVE(thread_data, entries);
+        free(thread_data);
+    }
+
+    // Close server socket and clean up
     if (server_socket != -1) close(server_socket);
     remove(FILE_PATH);
+    pthread_mutex_destroy(&file_mutex);
     closelog();
+
     exit(EXIT_SUCCESS);
 }
 
-void daemonize() {
-    pid_t pid, sid;
-    
-    pid = fork();
-    if (pid < 0) { exit(EXIT_FAILURE); }
-    if (pid > 0) { exit(EXIT_SUCCESS); }
+// Thread function to handle client communication
+void *client_handler(void *arg) {
+    client_thread_t *thread_data = (client_thread_t *)arg;
+    int client_socket = thread_data->client_socket;
+    char buffer[BUFFER_SIZE];
+    ssize_t bytes_received;
+
+    syslog(LOG_INFO, "Thread started for client socket: %d", client_socket);
+
+    FILE *file = fopen(FILE_PATH, "a+");
+    if (!file) {
+        syslog(LOG_ERR, "Failed to open file");
+        close(client_socket);
+        free(thread_data);
+        return NULL;
+    }
+
+    while ((bytes_received = recv(client_socket, buffer, BUFFER_SIZE, 0)) > 0) {
+        pthread_mutex_lock(&file_mutex);
+        fwrite(buffer, 1, bytes_received, file);
+        fflush(file);
+        pthread_mutex_unlock(&file_mutex);
+
+        if (memchr(buffer, '\n', bytes_received)) {
+            rewind(file);
+            while ((bytes_received = fread(buffer, 1, BUFFER_SIZE, file)) > 0) {
+                send(client_socket, buffer, bytes_received, 0);
+            }
+            break;
+        }
+    }
+
+    fclose(file);
+    close(client_socket);
+    syslog(LOG_INFO, "Thread finished for client socket: %d", client_socket);
 
-    sid = setsid();
-    if (sid < 0) { exit(EXIT_FAILURE); }
+    // Remove from list
+    LIST_REMOVE(thread_data, entries);
```

Comment:
What is the concern with doing LIST_REMOVE on the list from within the client thread?